drivers/serial/uart_nrf5e.c

```diff
@@ -0,0 +1,873 @@
+	} else {
+		/* We can read as many bytes as needed to
+		 * make read index equal to write index
+		 */
+		tx_size = dev_data->fifo_tx_write_index -
+			  dev_data->fifo_tx_read_index;
+	}
+
+	/* Limit the blocklength */
+	if (tx_size > CONFIG_UART_NRF5E_TX_DMA_MAXBLOCK) {
+		tx_size = CONFIG_UART_NRF5E_TX_DMA_MAXBLOCK;
+	}
+
+	/* And go */
+	uart->TXDMAXCNT = tx_size;
+	uart->TASKS_STARTTX = 1;
+}
+
+/** Interrupt driven FIFO fill function */
+static int uart_nrf5e_fifo_fill(struct device *dev,
+			 const u8_t *tx_data, int len)
+{
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+	u8_t num_tx = 0;
+
+	if (!tx_data) {
+		return 0;
+	}
+
+	uart_nrf5e_mask(dev, 1);
+
+	while (len - num_tx) {
+		u8_t new_index = dev_data->fifo_tx_write_index + 1;
+
+		if (new_index >= sizeof(dev_data->fifo_tx_data)) {
+			new_index = 0;
+		}
+
+		/* Is the fifo full */
+		if (new_index == dev_data->fifo_tx_read_index) {
+			break;
+		}
+
+		dev_data->fifo_tx_data[dev_data->fifo_tx_write_index] =
+								*tx_data;
+		dev_data->fifo_tx_write_index = new_index;
+
+		tx_data++;
+		num_tx++;
+	}
+
+	if (num_tx) {
+		uart_nrf5e_fifo_tx_start(dev);
+	}
+
+	uart_nrf5e_mask(dev, 0);
+
+	return (int)num_tx;
+}
+
+/** Interrupt driven FIFO read function */
+static int uart_nrf5e_fifo_read(struct device *dev, u8_t *rx_data, int size)
+{
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+	u8_t num_rx = 0;
+
+	uart_nrf5e_mask(dev, 1);
+
+	while (size - num_rx) {
+		if (dev_data->fifo_rx_write_index ==
+				dev_data->fifo_rx_read_index) {
+			break;
+		}
+
+		if (rx_data) {
+			*rx_data = dev_data->fifo_rx_data
+				[dev_data->fifo_rx_read_index];
+			rx_data++;
+		}
+
+		num_rx++;
+
+		u8_t tmp = dev_data->fifo_rx_read_index + 1;
+
+		if (tmp >= sizeof(dev_data->fifo_rx_data)) {
+			tmp = 0;
+		}
+
+		dev_data->fifo_rx_read_index = tmp;
+	}
+
+	if (num_rx) {
+		uart_nrf5e_fifo_rx_setup(dev);
+	}
+
+	uart_nrf5e_mask(dev, 0);
+
+	return num_rx;
+}
+
+
+static void uart_nrf5e_fifo_rx_setup(struct device *dev)
+{
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+	volatile struct _uart *uart = UART_STRUCT(dev);
+	u8_t len = 0;
+
+	if (dev_data->fifo_rx_operating) {
+		return;
+	}
+
+	/* How many bytes can we write? */
+	if (dev_data->fifo_rx_write_index < dev_data->fifo_rx_read_index) {
+		len = dev_data->fifo_rx_read_index -
+			dev_data->fifo_rx_write_index - 1;
+	} else {
+		len = sizeof(dev_data->fifo_rx_data) -
+			dev_data->fifo_rx_write_index;
+		if (dev_data->fifo_rx_read_index == 0) {
+			len--;
+		}
+	}
+
+	if (!len) {
+		return;
+	}
+
+	if (len > CONFIG_UART_NRF5E_RX_DMA_MAXBLOCK) {
+		len = CONFIG_UART_NRF5E_RX_DMA_MAXBLOCK;
+	}
+
+	dev_data->fifo_rx_operating = 1;
+
+	uart->RXD = (u32_t)&dev_data->fifo_rx_data
+			[dev_data->fifo_rx_write_index];
+	uart->RXDMAXCNT = len;
+	uart->TASKS_STARTRX = 1;
+}
+
+/** Interrupt driven transfer ready function */
+static int uart_nrf5e_irq_tx_ready(struct device *dev)
```

Comment:
According to my understanding, this function returns 0 if the TX FIFO queue is full. What is the reason that the function name is, instead, _irq_tx_ready?